### ISSUE TITLE
Connector health uses `master` not published version

### ISSUE THREAD

**Issue by evantahler** [airbytehq/airbyte#20442]

**Comment by evantahler** [comment]
cc @sherifnada and @YowanR - do you have any context to why the current process (building and testing the connector each time) is desirable to testing the published version of the connector?  Do we want to make a change here, or do both?

**Comment by sherifnada** [comment]
@evantahler what's the value in repeatedly running the build against the published docker version? i.e: why should we expect that build output to be different than whenever we ran it initially?

**Comment by evantahler** [comment]
That's the big question!  Here's what I think the options break down to:

* Testing published docker images ensures:
  * No API changes or changes in the source response payloads break our connectors or violate the schema
* Testing `master`
  * Ensures that the connector can be published again easily and currently works with the latest CDK and Airbyte Protocol
  * No API changes or changes in the source response payloads break our connectors or violate the schema

I think the fact that we are asking GL to ensure that these connector tests remain green implies to me that this is a form of customer support / OC prevention.  In that realm, I think we should be testing what the users actually would be using, not that our next build could be green.  I've seen some issues pop up in `#connector-health` which are cdk-update-related, and I'm glad we found them, but perhaps they don't need to be as high-priority if the published connector is OK. 

Tagging in @airbytehq/connector-operations for their thoughts as well - this topic came up in our latest planning meeting.

**Comment by erohmensing** [comment]
> what's the value in repeatedly running the build against the published docker version? i.e: why should we expect that build output to be different than whenever we ran it initially?

The reason this stuck out to me is that @YowanR seemed frustrated that despite his team doing active work fixing failures for integration tests, it didn't seem to be making lots of headway. I noticed when I pushed a new CDK and immediately new ones were failing - of course they can't make headway when we're introducing things breaking it every day 😅

If the goal of the #connector-health check is to make sure things are passing in production, then I think it makes sense to test the published versions, but I see what you mean about continually testing them.... maybe this could be a separate KPI we track and always expect to be at 100% unless something strange like API changes that evan mentioned? 

Having the tests running on master all the time is helpful for knowing when something you introduced might have had edge cases you might not have considered. But putting team goals around keeping these tests green other teams' work is causing the dependencies' code to change every day is.... complicated for sure

**Comment by alafanechere** [comment]
From a reliability standpoint, I would love to have the confidence that all connectors of our catalog are passing tests. This is not information we have with `master` nightly builds. 
Having a 🔴 test on a published connector version would help prioritize fixes and prevent OC issues.
Having this historized per version would also help boost rollback confidence.
Another approach would be to disallow merge on test failure. 
In this scenario, failing builds on the master can easily be assigned to dependency issue (CDK, SAT).
IMO for our peace of mind we should indeed have:
- the `master` nightly builds we have today
- A nightly `/test` run on the most recent version of each connector.

**Comment by evantahler** [comment]
Grooming: 
We want to test both `master` and the published connector, and share this info with GL so they can make priority choices about what to fix.

TODO: 
* decide if testing both happens nightly or some other frequency

**Comment by YowanR** [comment]
Catching up on this thread now. I agree with the conclusion above: we should test both master and the published connector. From a customer POV, is it correct to infer that `master` is addressing the needs of our OSS users and that the published connector is addressing the needs of our Cloud users?
I'd argue that we should run both on a daily basis but treat published connectors issues as P0 and `master` issues as P1 when it comes to fixing (if/when we have to choose between fixing both as OSS can revert to the published version). Thoughts on this?
Also cc @bleonard as FYI

**Comment by erohmensing** [comment]
> From a customer POV, is it correct to infer that master is addressing the needs of our OSS users and that the published connector is addressing the needs of our Cloud users?

I don't think this is exactly the case, because of the following from the description: 

>  [When we test on master], We are testing the connector with the /latest/ CDK, not what it was published with.

So it doesn't match the behavior that the open source users would get by using the `latest` published dokcer image of the connector, if the CDK has been updated since that latest version was published. 

> the published connector is addressing the needs of our Cloud users?

I guess actually checking the version in Cloud would be the way to 100% address the needs of cloud users, instead of just the latest published version.

**Comment by sherifnada** [comment]
thanks for the context y'all! running tests against published versions sounds valuable for determining that published versions work against the API. But I think not building against what's on master means that we're effectively no longer maintaining a "build status" for connectors in the same way that e.g: the platform maintains a build status for its master branch regardless of whether. Is that really the desired effect here? i.e: are we saying we do not care that connectors are always in a publishable state?

**Comment by evantahler** [comment]
From our latest grooming:

> We want to test both master and the published connector, and share this info with GL so they can make priority choices about what to fix.


So we want to continue the test we have today _and_ add an additional test for the published connector.  We may tweak the frequencies as we think about this more seriously

**Comment by bnchrch** [comment]
This issue was referenced in [Tech Spec: Connector Metadata and Build Pipeline](https://docs.google.com/document/d/1CXG0hTftomCTO3vB0ocbs00CkYBD6YURfrxm8KwozB4/edit#)

This is because there is a plan to rewrite the nightly report system.

@YowanR for context sake what of the two outcomes outlined above is more useful to the GL team
1. Knowing which connectors are broken on master
2. Knowing which connectors are broken in cloud

**Comment by YowanR** [comment]
@bnchrch I missed this earlier. Cloud is probably more important as these might lead to on-call issues but ideally, we would want both.

**Comment by erohmensing** [comment]
+1 on both - broken on cloud tells us what's happening to end users, but broken on master means decrease in dev speed since devs will run into failing tests when trying to publish new updates
